src/librbd/io/QosImageDispatch.cc

```diff
@@ -302,11 +308,15 @@ bool QosImageDispatch<I>::needs_throttle(
                       Tag{image_dispatch_flags, on_dispatched}, flag)) {
       ldout(cct, 15) << "on_dispatched=" << on_dispatched << ", "
                      << "flag=" << flag << dendl;
-      all_qos_flags_set = false;
     } else {
-      all_qos_flags_set = set_throttle_flag(image_dispatch_flags, flag);
+      flags_to_set |= flag;
     }
   }
+
+  if (flags_to_set != 0) {
+    all_qos_flags_set = set_throttle_flags(image_dispatch_flags, flags_to_set);
+  }
+
   return !all_qos_flags_set;
```

Comment:
`flags_to_set` can't really be 0 here because a single op can't activate all throttles -- see how `calculate_tokens()` returns 0 in some cases.  Another theoretical possibility is `m_throttles` being empty, but then we are in trouble anyway because the dispatch would hang indefinitely.  I'd suggest something like
```suggestion
  ceph_assert(flags_to_set != 0);
  return !set_throttle_flags(image_dispatch_flags, flags_to_set);
```
and getting rid of `all_qos_flags_set` entirely, not just of the assignment in the branch where throttle is activated.